### ISSUE TITLE
Notification Handler Type Registered Multiple Times

### ISSUE THREAD

**Issue by rthm93** [jbogard/MediatR.Extensions.Microsoft.DependencyInjection#36]
As per previously discussed in the following issue:
https://github.com/jbogard/MediatR/issues/262

Registering multiple notification handler types from the same assembly for the same notification will result to multiple instance of same notification handler to be instantiated causing same event to be handled multiple times.

Maybe we should add a check to find are the current service type and implementation type already registered into the collection to prevent duplicate registering?

**Comment by jbogard** [comment]
I assumed that TryXyz would only register. Is it possible to get a repro?

**Comment by rthm93** [comment]
Please refer to this [gist](https://gist.github.com/rthm93/492f68d9b041a34c0be3051e6cea65f8)

**Comment by jbogard** [comment]
I didn't have a few of those classes or interfaces defined, so I created a test to see if I could repro. I can't:

```c#
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Xunit;
using Xunit.Abstractions;

namespace MediatR.Extensions.Microsoft.DependencyInjection.Tests
{

    public class NotificationBugs
    {
        private readonly ITestOutputHelper _output;

        public class DomainEvent : INotification
        {
        }

        public abstract class BaseNotificationHandler<TDomainEvent> : INotificationHandler<TDomainEvent>
            where TDomainEvent : DomainEvent
        {
            /// <summary>
            /// Handles domain event.
            /// </summary>
            /// <param name="domainEvent">Domain event.</param>
            /// <param name="cancellationToken">Cancellation token.</param>
            public async Task Handle(TDomainEvent domainEvent, CancellationToken cancellationToken)
            {
                await HandleNotification(domainEvent, cancellationToken);
            }

            /// <summary>
            /// Actual implementation of notification handling.
            /// </summary>
            /// <param name="notification">Notification.</param>
            /// <param name="cancellationToken">Cancellation token.</param>
            protected abstract Task HandleNotification(TDomainEvent notification, CancellationToken cancellationToken);
        }

        public class SetDemoDataEvent : DomainEvent
        {
            public string Message { get; set; }
        }

        public class DemoEventHandler : BaseNotificationHandler<SetDemoDataEvent>
        {
            private readonly ITestOutputHelper _output;

            public DemoEventHandler(ITestOutputHelper output)
            {
                _output = output;
            }
            protected override Task HandleNotification(SetDemoDataEvent notification,
                CancellationToken cancellationToken)
            {
                _output.WriteLine($"Notified: {notification.Message}");

                return Task.CompletedTask;
            }
        }

        public NotificationBugs(ITestOutputHelper output) => _output = output;

        [Fact]
        public void ShouldNotify()
        {
            var services = new ServiceCollection();
            services.AddMediatR(typeof(NotificationBugs));
            services.AddSingleton(_output);
            var provider = services.BuildServiceProvider();

            var e = new SetDemoDataEvent {Message = "Hello"};

            using (var scope = provider.CreateScope())
            {
                var mediator = scope.ServiceProvider.GetService<IMediator>();
                mediator.Publish(e);
            }
        }

    }

}
```

**Comment by rthm93** [comment]
Please refer to this [gist](https://gist.github.com/rthm93/ea8a43b2fd0cd142d71ed6eb4dc0f5c9).
This should be able to reproduce it.

It happens when i registered multiple types from the same assembly.

And I noticed the parameter was called **`handlerAssemblyMarkerTypes`**
Were we supposed to pass in at most 1 type from each assembly? 

Example: 
In assembly 1 I have `XXXHandler` and `YYYHandler`
In assembly 2 I have `ZZZHandler`

What I should have done should be the following? 

`services.AddMediatR(new Type[] { typeof(XXXHandler), typeof(ZZZHandler) } );`

**Comment by jbogard** [comment]
Ahhhh OK, yes, the types you pass in aren't the handler types. They're just a shortcut to finding assemblies. This package scans your assemblies.

But I think that's an easy check on my side - to only pull out the distinct assemblies.

**Comment by rosdi** [comment]
I assume this has not been released in the latest nuget yet? I am registering my handles like so:

`services.AddMediatR();`
`services.AddMediatR(typeof(My.App.NameSpace.OneOfHandler).Assembly);`

And I am seeing all my handlers are being instantiated twice... I purposely created 3 handlers that handle the same notification, all of the 3 handlers are instantiated twice.

**Comment by jbogard** [comment]
It has, it's version 5.0.2 of the extensions package, release a few days
ago.

**Comment by rosdi** [comment]
Hmm... I am already using `MediatR.Extensions.MIcrosoft.DependencyInjection 5.0.2`, but for some reason my handlers still  being instantiated twice...

I will try reproduce this in a barebone ASP.NET Core project and reopen a new issue if it still happens.

**Comment by rosdi** [comment]
Seems like I am registering the handlers twice somehow..., I need to call `.AddMediatR` only once.

This works (handles will be instantiated only one, everything works as expected):
`//services.AddMediatR();`
`services.AddMediatR(typeof(My.App.NameSpace.OneOfHandler).Assembly);`

This does not work (nothing will happen, no handlers will be invoked):
`services.AddMediatR();`
`//services.AddMediatR(typeof(My.App.NameSpace.OneOfHandler).Assembly);`

This does not work (handles will be instantiated twice, each handlers will run twice for every message):
`services.AddMediatR();`
`services.AddMediatR(typeof(My.App.NameSpace.OneOfHandler).Assembly);`

Obviously... I do not understand something very obvious here.

**Comment by jbogard** [comment]
What, you were calling it twice? Why?

Just call it once, passing in the assemblies/marker types for the
assemblies you want.

This is like calling services.AddMvc() twice.

**Comment by hgmauri** [comment]
Any solution on multiple handlers?

I'm using the configuration below in **startup.cs** but I keep getting multiple handlers in INotificationHandler.

```csharp
private static IMediator BuildMediator(IServiceCollection services)
{
         services.AddScoped<ServiceFactory>(p => p.GetService);

         services.Scan(scan => scan
                .FromAssembliesOf(typeof(IMediator),
                    typeof(ClientCommand))
                .AddClasses()
                .AsImplementedInterfaces());

         var provider = services.BuildServiceProvider();

         return provider.GetRequiredService<IMediator>();
}

**Comment by bdaniel7** [comment]
I'm using version 7 of MediatR and MediatR.Extensions.DependencyInjection in an ASP .NET Core project.
The INotificationHandlers are still registered and executed twice.

I have this [gist](https://gist.github.com/bdaniel7/ee07f1278500febc160fec7263f0477a):

**Comment by rthm93** [comment]
I am not familiar with Autofac, but it seems that you are registering twice.  
`services.AddMediatR(domainAssembly);`   
This line registers into the default service collection.

`builder.RegisterAssemblyTypes(domainAssembly)`
This line registers into Autofac.

Maybe you could refer to [this](https://github.com/jbogard/MediatR/blob/master/samples/MediatR.Examples.Autofac/Program.cs)?

**Comment by bdaniel7** [comment]
Yes, I suspected that the handlers are registered twice, but I was using automatic registration, and I assumed it's some kind of misconfiguration.

**Comment by kerams** [comment]
Would it be possible to create an overload of `AddMediatR` that would only register the base types and handlers if they are not already present (calling it again with different assemblies should just add new handlers)? We have a RabbitMQ and EventStoreDB libraries that internally use MediatR to delegate messages/events to user created handlers. Both of these call `AddMediatR` in their respective `Add` methods. When you use both libraries in a single application, handlers get invoked twice. The workaround here is to require each application to call `AddMediatR` on its own if any of the libraries need it. It's not a huge deal, but it would be great if I did not have to worry about it.
